### PR TITLE
[Caffe2] Resolve error C2375 "redefinition; different linkage" by adding dllimport.

### PR DESCRIPTION
**Pull request by xkszltl**
For #10573

### CODE REVIEW

File: caffe2/core/operator_schema.h [hunk 1]
```diff
@@ -324,7 +324,7 @@ class CAFFE2_API OpSchema {
     return std::numeric_limits<int>::max();
   }
 
-  friend std::ostream& operator<<(std::ostream& out, const OpSchema& schema);
+  CAFFE2_IMPORT friend std::ostream& operator<<(std::ostream& out, const OpSchema& schema);
```
[the remark this answers is not in the document]
Comment (xkszltl) [reply caffe2/core/operator_schema.h]: Both should work and both are weird...But I agree maybe `extern dllexport` is a little bit less weird than overriding `dllimport` with `dllexport`...

Based on the test, seems cuda8 doesn't like this syntax

[the remark this answers is not in the document]
Comment (xkszltl) [reply caffe2/core/operator_schema.h]: FYI here's a doc for `extern dllexport`: https://msdn.microsoft.com/en-us/library/y4h7bcy6.aspx

**Comment by xkszltl** [comment]
Merge conflict resolved.

File: caffe2/core/registry.h [hunk 2]
```diff
@@ -152,7 +152,7 @@ class CAFFE2_API Registerer {
  */
 #define CAFFE_DECLARE_TYPED_REGISTRY(                                    \
     RegistryName, SrcType, ObjectType, PtrType, ...)                     \
-  CAFFE2_EXPORT Registry<SrcType, PtrType<ObjectType>, ##__VA_ARGS__>*   \
```
Comment (xkszltl) [inline caffe2/core/registry.h]: Any reason for CAFFE2_EXPORT a declaration? It'll lead to undefined reference when using it outside the main proj.

**Comment by ezyang** [comment]
This diff is obsolete now, all changes are applied on master.

**Pull request closed without merging**